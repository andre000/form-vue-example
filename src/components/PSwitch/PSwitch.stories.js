import '@/assets/scss/storybook.scss'
import PSwitch from './PSwitch.vue'

export default {
  component: PSwitch,
  title: 'PSwitch'
}

export const withInitialState = () => ({
  components: { PSwitch },
  props: {
    value: {
      default: 1
    }
  },
  template: `
    <div class="background">
      <p-switch v-model="value" :first-option="{ label: 'One', value: 1 }" :second-option="{ label: 'Two', value: 2 }"/>
    <div>
  `
})
