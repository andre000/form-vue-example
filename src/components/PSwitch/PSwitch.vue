<template>
  <label for="toggle-button" class="switch">
    <span :class="value === firstOption.value && 'selected'">{{
      firstOption.label
    }}</span>
    <input
      ref="toggle-button"
      type="checkbox"
      id="toggle-button"
      v-on="listeners"
      v-bind="$attrs"
      v-model="currentState"
      style="display: none"
    />

    <span class="switch__bar">
      <i class="switch__button" :class="buttonPosition"> </i>
    </span>
    <span :class="value === secondOption.value && 'selected'">{{
      secondOption.label
    }}</span>
  </label>
</template>

<script>
const optionValidator = option =>
  Object.keys(option).every(op => ['label', 'value'].includes(op)) &&
  Object.keys(option).length === 2

export default {
  name: 'PSwitch',
  props: {
    value: {
      type: [String, Number],
      required: true
    },
    firstOption: {
      type: Object,
      required: true,
      validator: optionValidator
    },
    secondOption: {
      type: Object,
      required: true,
      validator: optionValidator
    }
  },

  data: () => ({
    currentState: false
  }),

  mounted() {
    if (this.value === this.secondOption.value) {
      this.currentState = true
    }
  },

  computed: {
    listeners() {
      return {
        ...this.$listeners,
        input: this.updateValue
      }
    },

    buttonPosition() {
      return this.value === this.firstOption.value ? 'left' : 'right'
    }
  },

  methods: {
    updateValue() {
      const value = this.currentState
        ? this.firstOption.value
        : this.secondOption.value
      this.$emit('input', value)
    }
  }
}
</script>

<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';

.switch {
  font: 400 16px/21px Roboto;
  color: $primary-text;
  cursor: pointer;

  .selected {
    font-weight: bold;
  }

  .switch__bar {
    margin: 0px 5px;
    background: #fff;
    width: 42px;
    height: 12px;
    display: inline-block;
    position: relative;
    border-radius: 11px;

    .switch__button {
      background: #e33535 0% 0% no-repeat padding-box;
      width: 20px;
      height: 20px;
      display: inline-block;
      position: absolute;
      border-radius: 100%;
      bottom: -4px;
      transition: transform 0.2s ease-in-out;

      &.left {
        transform: translateX(0px);
      }
      &.right {
        transform: translateX(calc(100% + 2px));
      }
    }
  }
}
</style>
