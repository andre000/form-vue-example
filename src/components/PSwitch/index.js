export { default as PSwitch } from './PSwitch'
