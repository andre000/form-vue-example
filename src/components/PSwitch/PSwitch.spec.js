import { shallowMount, createLocalVue } from '@vue/test-utils'

import PSwitch from './PSwitch'

const localVue = createLocalVue()

const mockedSwitch = {
  value: 0,
  firstOption: { label: 'Foo', value: 0 },
  secondOption: { label: 'Bar', value: 1 }
}

describe('PSwitch component', () => {
  let wrapper
  beforeAll(() => {
    wrapper = shallowMount(PSwitch, {
      localVue,
      propsData: { ...mockedSwitch }
    })
  })

  it('should match the snapshot', () =>
    expect(wrapper.html()).toMatchSnapshot())

  it('should have both options label', () => {
    expect(wrapper.find('span:first-child').text()).toMatch(
      mockedSwitch.firstOption.label
    )
    expect(wrapper.find('span:last-child').text()).toMatch(
      mockedSwitch.secondOption.label
    )
  })

  it('should have a hidden checkbox input', () => {
    const input = wrapper.find('input[type="checkbox"]')
    expect(input.exists()).toBe(true)
    expect(global.getComputedStyle(input.element).display).toBe('none')
  })

  it('should change values when clicked', async () => {
    await wrapper.trigger('click')
    expect(wrapper.vm.currentState).toBe(true)
    expect(wrapper.emitted().input[0][0]).toBe(mockedSwitch.secondOption.value)
  })

  it('should change switch state with initial value prop', async () => {
    const wrapperNewValue = shallowMount(PSwitch, {
      localVue,
      propsData: { ...mockedSwitch, value: 1 }
    })
    expect(wrapperNewValue.vm.currentState).toBe(true)
  })
})
