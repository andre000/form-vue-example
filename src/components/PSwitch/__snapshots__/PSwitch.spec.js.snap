// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PSwitch component should match the snapshot 1`] = `<label for="toggle-button" class="switch"><span class="selected">Foo</span> <input type="checkbox" id="toggle-button" style="display: none;"> <span class="switch__bar"><i class="switch__button left"></i></span> <span class="">Bar</span></label>`;
